get bond interface details and save cfg

vaishnavibhat opened the pull request:
Use sysfs param to get the bond interface details and
save the master and slave cfg file

Signed-off-by: Vaishnavi Bhat <vaishnavi@linux.vnet.ibm.com>

vaishnavibhat commented:
Patch to create cfg file in the format :

Master cfg
------------------
TYPE=Bond
BOOTPROTO=static
NAME=bondtest
DEVICE=bondtest
ONBOOT=yes
IPADDR=156.1.1.173
NETMASK=255.255.255.0
IPV6INIT=yes
IPV6_AUTOCONF=yes
IPV6_DEFROUTE=yes
BONDING_MASTER=yes
BONDING_OPTS=mode=active-backup

Slave cfg
-------------
SLAVE=yes
ONBOOT=yes
MASTER=bondtest
NAME=net1
DEVICE=net1
TYPE=Ethernet

clebergnu commented on avocado/utils/network/interfaces.py:
```diff
@@ -75,6 +75,11 @@ def _get_interface_details(self, version=4):
             log.error(msg)
             raise NWException(msg)
 
+    def _get_bondinterface_details(self, version=4):
```
The `version` parameter is not used here.

clebergnu replied:
@vaishnavibhat if my comment was accurate, and `version` was not used on that version, my suggestion was in the lines of removing the parameter.  I see that the current version *does* use `version`, so I'd go with a better default such as you had before (and matches `getipaddrs`).

clebergnu replied:
@vaishnavibhat I apologize if my previous comment was mistaken and led to rework.

beraldoleal replied:
IIUC, this has still some work to do here, since version is not being used on this method.

clebergnu commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +318,36 @@ def save(self, ipaddr, netmask):
+
+        if self.if_type == 'Bond':
+            # Update master and slave cfg for bond
+            ifcfg_dict.update({'BONDING_MASTER': 'yes'})
```
`update` is usually a good idea if you have more than one variable to set.  In this case, I find the following preferrable:

```python
ifcfg_dict['BONDING_MASTER'] = 'yes'
```

clebergnu commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +318,36 @@ def save(self, ipaddr, netmask):
+                                'ONBOOT': 'yes',
+                                'MASTER': self.name}
+            if current_distro.name == 'SuSE':
+                ifcfg_dict.update({'BONDING_MODULE_OPTS=mode': mode})
```
This looks weird to me... don't you mean:

```
ifcfg_dict.update({'BONDING_MODULE_OPTS': mode})
```

Or, according to the previous suggestion:

```python
ifcfg_dict['BONDING_MODULE_OPTS'] = mode
```

clebergnu commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +318,36 @@ def save(self, ipaddr, netmask):
+                                             'DEVICE': slave})
+                    self._write_to_file("{}/ifcfg-{}".format(path, slave), ifcfg_slave_dict)
+            elif current_distro.name in ['rhel', 'fedora']:
+                ifcfg_dict.update({'BONDING_OPTS=mode': mode})
```
Likewise.

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -75,6 +75,11 @@ def _get_interface_details(self, version=4):
+    def _get_bondinterface_details(self, version=4):
+        cmd = "cat /sys/class/net/{}/bonding/mode /sys/class/net/{}/bonding/slaves".format(self.name, self.name)
+        output = run_command(cmd, self.host)
+        return output
```
Some interfaces don't have bonding directory. We need to handle here,  and return None instead of an exception.

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +318,36 @@ def save(self, ipaddr, netmask):
+            # Update master and slave cfg for bond
+            ifcfg_dict.update({'BONDING_MASTER': 'yes'})
+            mode = self._get_bondinterface_details().splitlines()[0].split()[0]
+            slaves = self._get_bondinterface_details().splitlines()[1].split()
```
You are calling the same command twice. One suggestion is to change the method so it will return a tuple.

```python
mode, slaves = self._get_bondinterface_details()
```

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +318,36 @@ def save(self, ipaddr, netmask):
+            if current_distro.name == 'SuSE':
+                ifcfg_dict.update({'BONDING_MODULE_OPTS=mode': mode})
+                for j in range(0, len(slaves)):
+                    BONDING_SLAVE = 'BONDING_SLAVE{}'.format(j)
```
s/BONDING_SLAVE/bonding_slave/g

ana commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +321,35 @@ def save(self, ipaddr, netmask):
             ifcfg_dict.pop('BOOTPROTO')
+
+        if self.if_type == 'Bond':
+            # Update master and slave cfg for bond
```
If you change "cfg" for "configuration", the spellchecker check will pass.

ana commented on avocado/utils/network/interfaces.py:
```diff
@@ -75,6 +75,14 @@ def _get_interface_details(self, version=4):
+            output = run_command(cmd, self.host)
+            return output
+        except Exception as ex:
+            raise NWException("Bond interface details not found")
```
You raise the exception as "ex" and never use this variable.

beraldoleal replied:
Would be nice instead of hiding the exception to exporting it to upper layers, since this is an generic exception and could get another problems.

```python
except Exception as ex:
   raise NWException("Bond interface details not found: {}".format(str(ex))
```

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -75,6 +75,14 @@ def _get_interface_details(self, version=4):
+        cmd = "cat /sys/class/net/{}/bonding/mode /sys/class/net/{}/bonding/slaves".format(self.name, self.name)
+        try:
+            output = run_command(cmd, self.host)
+            return output
```
Don't need variable here.

```python
return run_command(cmd, self.host)
```

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +321,35 @@ def save(self, ipaddr, netmask):
+                                'MASTER': self.name}
+            if current_distro.name == 'SuSE':
+                ifcfg_dict['BONDING_MODULE_OPTS'] = 'mode='+mode.split()[0]
+                for j in range(0, len(slaves.split())):
```
Your _get_bondinterface_details(), could return a list on slaves instead of a string, so here would be easier for you:

```python
for slave in slaves:
```

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +321,35 @@ def save(self, ipaddr, netmask):
+                ifcfg_dict['BONDING_MODULE_OPTS'] = 'mode='+mode.split()[0]
+                for j in range(0, len(slaves.split())):
+                    bonding_slave = 'BONDING_SLAVE{}'.format(j)
+                    slave = slaves.split()[j]
```
and that line will not be needed.

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +321,35 @@ def save(self, ipaddr, netmask):
+            elif current_distro.name in ['rhel', 'fedora']:
+                ifcfg_dict['BONDING_OPTS'] = 'mode='+mode.split()[0]
+                for j in range(0, len(slaves.split())):
+                    slave = slaves.split()[j]
```
same here.

beraldoleal reviewed (changes requested):
Hi @vaishnavibhat, thanks for you patience here. I just left a few more comments. We are close to a release, let me know if you are going to address those, otherwise I can send a new version.

vaishnavibhat commented:
@beraldoleal addressed the above comments. Can you please confirm if it looks okay ?

ana commented:
@vaishnavibhat There were a lot of changes on avocado and even a release in the last weeks. Before reviewing again, could you please rebase your changes on top of the master branch? Thanks

vaishnavibhat commented:
Hi @ana Does the latest patch look fine ? Please suggest. 

Thanks

clebergnu reviewed (changes requested):
Hi @vaishnavibhat , it looks like you addressed all of the reviewers' comments.  I left another comment about the `version` parameter.  I *thought* it was not used on the version I initially looked at, but now I'm not that sure anymore.

Given that `version` is now used, I think the previous default `4` is a better choice indeed.

vaishnavibhat commented:
> Hi @vaishnavibhat , it looks like you addressed all of the reviewers' comments. I left another comment about the `version` parameter. I _thought_ it was not used on the version I initially looked at, but now I'm not that sure anymore.
> 
> Given that `version` is now used, I think the previous default `4` is a better choice indeed.

Included version=4 as def _get_bondinterface_details(self, version=4) in the latest patch.

vaishnavibhat commented:
@clebergnu Requesting a re-review. 
Thank you.

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -71,10 +73,17 @@ def _get_interface_details(self, version=4):
+    def _get_bondinterface_details(self, version=4):
+        cmd = "cat /sys/class/net/{}/bonding/mode /sys/class/net/{}/bonding/slaves".format(self.name, self.name)
+        try:
+            return run_command(cmd, self.host)
```
Here you are returning the raw output, leaving to upper layer methods to handle this output. Would be nice to return here the specific information or a dict so caller methods wouldn't need to do splitlines(), split() and stuff like that.

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +322,35 @@ def save(self, ipaddr, netmask):
+        if self.if_type == 'Bond':
+            # Update master and slave configuration for bond
+            ifcfg_dict['BONDING_MASTER'] = 'yes'
+            mode, slaves = self._get_bondinterface_details().splitlines()
```
This is what I'm talking about.

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +322,35 @@ def save(self, ipaddr, netmask):
+                                'ONBOOT': 'yes',
+                                'MASTER': self.name}
+            if current_distro.name == 'SuSE':
+                ifcfg_dict['BONDING_MODULE_OPTS'] = 'mode='+mode.split()[0]
```
And here.

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +322,35 @@ def save(self, ipaddr, netmask):
+                                'MASTER': self.name}
+            if current_distro.name == 'SuSE':
+                ifcfg_dict['BONDING_MODULE_OPTS'] = 'mode='+mode.split()[0]
+                for j in range(0, len(slaves.split())):
```
And here.

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +322,35 @@ def save(self, ipaddr, netmask):
+                                             'DEVICE': slave})
+                    self._write_to_file("{}/ifcfg-{}".format(path, slave), ifcfg_slave_dict)
+            elif current_distro.name in ['rhel', 'fedora']:
+                ifcfg_dict['BONDING_OPTS'] = 'mode='+mode.split()[0]
```
And here.

beraldoleal commented on avocado/utils/network/interfaces.py:
```diff
@@ -313,6 +322,35 @@ def save(self, ipaddr, netmask):
+                    self._write_to_file("{}/ifcfg-{}".format(path, slave), ifcfg_slave_dict)
+            elif current_distro.name in ['rhel', 'fedora']:
+                ifcfg_dict['BONDING_OPTS'] = 'mode='+mode.split()[0]
+                for j in range(0, len(slaves.split())):
```
And here.

beraldoleal commented:
Also,  spell and lint are failing:
```

check-951-selftests/spell.sh FAIL
check-957-selftests/lint.sh FAIL
```

vaishnavibhat commented:
@beraldoleal @clebergnu  Any updates on this PR ?

clebergnu reviewed (approved):
LGTM, thanks!
